mls-rs/src/group/mod.rs

```diff
@@ -1617,6 +1548,145 @@ where
     }
 }
 
+impl<C: ClientConfig> Group<C> {
+    #[cfg(feature = "psk")]
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    async fn psk_secret<CS: CipherSuiteProvider>(
+        config: &C,
+        cipher_suite_provider: &CS,
+        psks: &[PreSharedKeyID],
+        additional_psk: Option<PskSecretInput>,
+    ) -> Result<PskSecret, MlsError> {
+        if let Some(psk) = additional_psk {
+            let psk_id = psks.first().ok_or(MlsError::UnexpectedPskId)?;
+
+            match &psk_id.key_id {
+                JustPreSharedKeyID::Resumption(r) if r.usage != ResumptionPSKUsage::Application => {
+                    Ok(())
+                }
+                _ => Err(MlsError::UnexpectedPskId),
+            }?;
+
+            let mut psk = psk;
+            psk.id.psk_nonce = psk_id.psk_nonce.clone();
+            PskSecret::calculate(&[psk], cipher_suite_provider).await
+        } else {
+            PskResolver::<
+                <C as ClientConfig>::GroupStateStorage,
+                <C as ClientConfig>::KeyPackageRepository,
+                <C as ClientConfig>::PskStore,
+            > {
+                group_context: None,
+                current_epoch: None,
+                prior_epochs: None,
+                psk_store: &config.secret_store(),
+            }
+            .resolve_to_secret(psks, cipher_suite_provider)
+            .await
+        }
+    }
+
+    #[cfg(not(feature = "psk"))]
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    async fn psk_secret<CS: CipherSuiteProvider>(
+        _config: &C,
+        cipher_suite_provider: &CS,
+        _psks: &[PreSharedKeyID],
+    ) -> Result<PskSecret, MlsError> {
+        Ok(PskSecret::new(cipher_suite_provider))
+    }
```

Comment:
I feel like having all this logic in group is a bad thing and we should plan on how to move it someplace more modular?

Reply:
Yeah agreed